### PR TITLE
Add SetModels to jujuclient API.

### PR DESCRIPTION
**Pull request by anastasiamac**
## Description of change

We need means of updating current stored controller models in one go.
This PR provides this functionality by introducing new jujuclient API to SetModels. This call essentially replaces stored models for controller with a new set. Only file implementation of ClientStore is tested extensively for this method in this PR since in-memory implementation is only used in commands testing and will be under scrutiny in the PRs that will update commands and their respective testing.  

This PR also removes old code with a TODO. This code was responsible for ensuring upgradability between 2.0 betas.

In addition, an amalgamated ValidateModel is provided for simplicity. This method validates both model name and details in one go as the same logic is needed in several places.

## QA steps

As this is an internal change of jujuclient API, all unit tests pass. There is no means of driving this functionality yet until it is plugged.

## Documentation changes
n/a

## Bug reference
n/a

### CODE REVIEW

File: jujuclient/file.go [hunk 1]
```diff
@@ -518,10 +515,9 @@ func (s *store) RemoveModel(controllerName, modelName string) error {
 	))
 }
 
-func updateModels(
-	controllerName string,
-	update func(*ControllerModels) (bool, error),
-) error {
+type modelUpdaterF func(storedModels *ControllerModels) (bool, error)
```
Comment (axw) [inline jujuclient/file.go]: conventionally it's "Func", not "F"
I think a more typical naming would be "updateModelFunc"

File: jujuclient/setmodels_test.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+func (s *ControllerModelsSuite) TestControllerModelsUpdate(c *gc.C) {
+	s.assertControllerStored(c)
+	expected := map[string]jujuclient.ModelDetails{
+		"admin/new-model":     s.assertModelStored(c, "admin/new-model", "test.model.uuid"),
+		"admin/another-model": s.assertModelStored(c, "admin/another-model", "test.model.uuid.2"),
+	}
+	s.assertStoreUpdated(c, expected)
+	s.assertStoreUpdated(c, expected)
```
Comment (axw) [inline jujuclient/setmodels_test.go]: delete duplicate

File: jujuclient/setmodels_test.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	s.FakeJujuXDGDataHomeSuite.TearDownTest(c)
+}
+
+func (s *ControllerModelsSuite) TestSetModelsNoController(c *gc.C) {
+	err := s.store.SetModels(s.controllerName, nil)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *ControllerModelsSuite) TestSetModelsNoControllerModels(c *gc.C) {
+	s.assertControllerStored(c)
```
Comment (axw) [inline jujuclient/setmodels_test.go]: seeing as all but one of the tests want a controller, how about inverting it so you add a controller in SetUpTest, and then remove it in TestSetModelsNoController?

also, please rename "assertControllerStored" so that it says what it does. it *adds* a controller. if you just move the logic to the SetUpTest, you can remove the method altogether

File: jujuclient/setmodels_test.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	err = s.store.SetModels("another-kontroller", otherModels)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// initial controller un-touched
+	storedModels, err := s.store.AllModels(s.controllerName)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(storedModels, gc.DeepEquals, before)
+}
+
+func (s *ControllerModelsSuite) assertStoreUpdated(c *gc.C, models map[string]jujuclient.ModelDetails) {
```
Comment (axw) [inline jujuclient/setmodels_test.go]: it's hard to follow the tests, because the method names don't do what I would think they do based on their names

s/assertStoreUpdated/assertSetModels/ please

File: jujuclient/setmodels_test.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	err := s.store.SetModels(s.controllerName, models)
+	c.Assert(err, jc.ErrorIsNil)
+	storedModels, err := s.store.AllModels(s.controllerName)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(storedModels, gc.DeepEquals, models)
+}
+
+func (s *ControllerModelsSuite) assertModelStored(c *gc.C, modelName, modelUUID string) jujuclient.ModelDetails {
```
Comment (axw) [inline jujuclient/setmodels_test.go]: and s/assertModelStored/assertUpdateModel/

File: jujuclient/setmodels_test.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+import (
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/jujuclient"
+	"github.com/juju/juju/testing"
+)
+
+type ControllerModelsSuite struct {
```
Comment (axw) [inline jujuclient/setmodels_test.go]: why create a whole new suite? there's ModelsSuite, which already has tests for UpdateModel. just add tests for SetModels to that?

File: jujuclient/utils.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+func updateControllerModels(storedModels *ControllerModels, newModels map[string]ModelDetails) (bool, error) {
+	if len(storedModels.Models) == 0 && len(newModels) == 0 {
+		// Special case: no update necessary.
+		return false, nil
+	}
+	// Add or update controller models based on a new collection.
+	for modelName, details := range newModels {
+		if err := ValidateModel(modelName, details); err != nil {
+			// TODO what to do here?... don't want to stop
```
Comment (axw) [inline jujuclient/utils.go]: yes, I think we do want to stop -- all or nothing

File: jujuclient/file.go [hunk 8]
```diff
@@ -547,6 +543,27 @@ func updateModels(
 	return nil
 }
 
+// SetModels implements ModelUpdater.
+func (s *store) SetModels(controllerName string, models map[string]ModelDetails) error {
+	if err := ValidateControllerName(controllerName); err != nil {
+		return errors.Trace(err)
+	}
+
```
Comment (axw) [inline jujuclient/file.go]: please validate models here, to avoid an unnecessary file read and lock below

File: jujuclient/mem.go [hunk 9]
```diff
@@ -183,20 +183,41 @@ func (c *MemStore) UpdateModel(controller, model string, details ModelDetails) e
...
+}
+
+// SetModels implements ModelUpdater.
+func (c *MemStore) SetModels(controller string, models map[string]ModelDetails) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if err := ValidateControllerName(controller); err != nil {
 		return err
 	}
```
Comment (axw) [inline jujuclient/mem.go]: please validate models here

File: jujuclient/utils.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+		}
+		storedModels.Models[modelName] = details
+	}
+	// Delete models that are not in the new collection.
+	for modelName, _ := range storedModels.Models {
+		if _, ok := newModels[modelName]; !ok {
+			delete(storedModels.Models, modelName)
+		}
+	}
+	return true, nil
```
Comment (axw) [inline jujuclient/utils.go]: the special case can be dropped by keeping a "changed" variable, making the code a little cleaner; and the function should return false if the model details don't change

```go
changed := len(storedModels.Models) != len(newModels)
for modelName := range storedModels.Models {
    if _, ok := newModels[modelName]; !ok {
        delete(storedModels.Models, modelName)
    }
}
for modelName, details := range newModels {
    oldDetails, ok := storedModels.Models[modelName]
    if ok && details == oldDetails {
        continue
    }
    storedModels.Models[modelName] = details
    changed = true
}
return changed, nil
```

please inline this method in the file store implementation's SetModels. The mem store implementation can just replace the map contents, since it's cheap

File: jujuclient/file.go (hunk 8 above)
Comment (anastasiamac) [inline jujuclient/file.go]: The validattion happens below inside updateModels. 
My thinking was that since I am updating a collection, it's best to check validity closer to the write. I am happy to move it out though - it's causing me puzzlement withing the update method :D I am pretty sure I've left a to-do because of it.

File: jujuclient/setmodels_test.go (hunk 6 above)
Comment (anastasiamac) [inline jujuclient/setmodels_test.go]: Yes, I could but it felt to me that existing suite had more than what I needed in terms of controllers and models.

File: jujuclient/file.go (hunk 8 above)
Comment (axw) [inline jujuclient/file.go]: Yep, just to be clear: it should happen before you get to updateModels, because updateModels requires that you take the lock and read the file. That sort of thing should be avoided if possible (and in this case, it is).

File: jujuclient/setmodels_test.go (hunk 2 above)
Comment (anastasiamac) [inline jujuclient/setmodels_test.go]: This was not an un-intentional duplicate. This was an additional call to ensure there is no changes to the data. I can see, however, that it's confusing. I might remove.

Comment (axw) [inline jujuclient/setmodels_test.go]: I made that comment before I understood that "assertStoreUpdated" made changes, and wasn't just asserting the state of the models. Probably best to leave a comment saying why you're doing it.

File: jujuclient/utils.go (hunk 7 above)
Comment (anastasiamac) [inline jujuclient/utils.go]: Well, this validation will move outside of the update. So there would not be an error here :D

File: jujuclient/utils.go (hunk 10 above)
Comment (anastasiamac) [inline jujuclient/utils.go]: Sounds good

File: jujuclient/file.go [hunk 11]
```diff
@@ -547,6 +544,57 @@ func updateModels(
...
+			changed = true
+		}
+		// Delete models that are not in the new collection.
+		for modelName, _ := range storedModels.Models {
+			if _, ok := models[modelName]; !ok {
+				delete(storedModels.Models, modelName)
+				changed = true
+			}
+		}
+		if storedModels.CurrentModel != "" && !modelNames.Contains(storedModels.CurrentModel) {
```
Comment (axw) [inline jujuclient/file.go]: get rid of "modelNames", and move this logic into the loop above

```go
// Delete models that are not in the new collection.
for modelName, _ := range storedModels.Models {
    if _, ok := models[modelName]; !ok {
        delete(storedModels.Models, modelName)
        changed = true
        if storedModels.CurrentModel == modelName {
            storedModels.CurrentModel = ""
        }
    }
}
```

File: jujuclient/mem.go [hunk 12]
```diff
@@ -200,6 +197,36 @@ func (c *MemStore) UpdateModel(controller, model string, details ModelDetails) e
...
+	}
+
+	controllerModels, ok := c.Models[controller]
+	if !ok {
+		controllerModels = &ControllerModels{
+			Models: make(map[string]ModelDetails),
+		}
+	}
+	controllerModels.Models = models
+	if controllerModels.CurrentModel != "" && !modelNames.Contains(controllerModels.CurrentModel) {
```
Comment (axw) [inline jujuclient/mem.go]:
```go
if _, ok := models[controllerModels.CurrentModel]; !ok {
    controllerModels.CurrentModel = ""
}
```

**Comment by anastasiamac** [comment]
!!build!!

check failed for external reasons:
Connection to ec2-34-229-192-129.compute-1.amazonaws.com closed by remote host.

File: jujuclient/file.go [hunk 13]
```diff
@@ -547,6 +544,57 @@ func updateModels(
...
+				continue
+			}
+			storedModels.Models[modelName] = details
+			changed = true
+		}
+		// Delete models that are not in the new collection.
+		for modelName, _ := range storedModels.Models {
+			if _, ok := models[modelName]; !ok {
+				delete(storedModels.Models, modelName)
+				changed = true
```
Comment (axw) [inline jujuclient/file.go]: changed = true is redundant here. if we got here, then either:

 - len(storedModels.Models) != len(models), in which case changed == true
 - len(storedModels.Models) == len(models), which implies that the same number of models were added as were removed

**Comment by Veebers** [comment]
!!build!!

File: jujuclient/file.go (hunk 11 above)
Comment (anastasiamac) [inline jujuclient/file.go]: nice :D done.

File: jujuclient/mem.go (hunk 12 above)
Comment (anastasiamac) [inline jujuclient/mem.go]: done

File: jujuclient/file.go (hunk 13 above)
Comment (anastasiamac) [inline jujuclient/file.go]: true, removed :D

**Comment by anastasiamac** [comment]
$$merge$$
